q2cli/cli.py

```diff
@@ -0,0 +1,113 @@
+# ----------------------------------------------------------------------------
+# Copyright (c) 2016--, QIIME 2 development team.
+#
+# Distributed under the terms of the Modified BSD License.
+#
+# The full license is in the file COPYING.txt, distributed with this software.
+# ----------------------------------------------------------------------------
+
+import click
+import qiime
+from qiime.sdk import PluginManager, SubprocessExecutor
+
+from . import __version__ as q2cli_version
+
+@click.group()
+@click.option('--wait/--dont-wait', default=True,
+              help="Wait/don't wait until process completes to return.",
+              show_default=True)
+@click.pass_context
+def cli(ctx, wait):
+    ctx.obj['plugin_manager'] = PluginManager()
+    ctx.obj['wait'] = wait
+
+@cli.command()
+def version():
+    click.echo('QIIME version: %s' % qiime.__version__)
+    click.echo('q2cli version: %s' % q2cli_version)
+
+@cli.command()
+@click.pass_context
+def plugins(ctx):
+    installed_plugins = ctx.obj['plugin_manager'].plugins
+    if len(installed_plugins) == 0:
+        click.echo('No plugins are currently installed.\nYou can browse the '
+                   'official QIIME 2 plugins at: https://github.com/qiime2.')
+    else:
+        click.echo('Installed plugins:')
+        for name, plugin in installed_plugins.items():
+            click.echo(' %s %s (%s)' %
+                       (name, plugin.version, plugin.website))
+
+# should the QIIME types be hashable to support this without casting
+# to a string?
+_type_map = {'Int': click.INT, 'Str': click.STRING, 'Float': click.FLOAT}
+
+def _create_callback(wf):
+    def f(ctx, **kwargs):
+        # execute workflow
+        executor = SubprocessExecutor()
+        input_artifacts = {
+            ia_name: kwargs[ia_name]
+            for ia_name in wf.signature.input_artifacts}
+        input_parameters = {
+            ip_name: kwargs[ip_name]
+            for ip_name in wf.signature.input_parameters}
+        output_artifacts = {
+            oa_name: kwargs[oa_name]
+            for oa_name in wf.signature.output_artifacts
+        }
+        future_ = executor(wf,
+                           input_artifacts,
+                           input_parameters,
+                           output_artifacts)
+
+        if ctx.obj['wait']:
+            # block until result is ready
+            completed_process = future_.result()
+
+            if completed_process.returncode != 0:
+                print(completed_process.stdout,
+                      file=sys.stdout)
+                print(completed_process.stderr,
+                      file=sys.stderr)
+                ctx.exit(1)
+        else:
+            # hmmm, I think this is still blocking
```

Comment:
#3 is updated with some details on this, we're going to come back to it. It's not trivial unfortunately.